src/gateway/sharding/shard_manager.rs

```diff
@@ -390,14 +379,70 @@ pub struct ShardManagerOptions {
     #[cfg(feature = "voice")]
     pub voice_manager: Option<Arc<dyn VoiceGatewayManager>>,
     pub ws_url: Arc<str>,
+    pub compression: TransportCompression,
     pub shard_total: NonZeroU16,
+    pub max_concurrency: NonZeroU16,
+    pub wait_time_between_shard_start: Duration,
     #[cfg(feature = "cache")]
     pub cache: Arc<Cache>,
     pub http: Arc<Http>,
     pub intents: GatewayIntents,
     pub presence: Option<PresenceData>,
-    pub max_concurrency: NonZeroU16,
-    /// Number of seconds to wait between starting each shard/set of shards start
-    pub wait_time_between_shard_start: Duration,
-    pub compression: TransportCompression,
+}
+
+pub enum ShardManagerMessage {
+    PollShardQueue,
+    Restart(ShardId),
+    Err(GatewayError),
+}
+
+/// A queue of [`ShardId`]s that is split up into multiple buckets according to the value of
+/// [`max_concurrency`](crate::model::gateway::SessionStartLimit::max_concurrency).
+#[must_use]
+pub struct ShardQueue {
+    buckets: FixedArray<VecDeque<ShardId>, u16>,
+    concurrent: bool,
+}
```

Comment:
I feel like ShardQueue should be split out into it's own file, because it's super easy to decouple and already seems a little out of place at the end of this file.

Reply:
Done